core/client/client.go

```diff
@@ -151,17 +164,18 @@ func min(x, y int32) int32 {
 	return y
 }
 
-// ListHistory returns a list of map roots and profiles at each revision.
-func (c *Client) ListHistory(ctx context.Context, userID, appID string, start, end int64) (map[uint64]*types.MapRootV1, map[uint64][]byte, error) {
+// PaginateHistory iteratively calls ListHistory to satisfy the start and end requirements.
+// Returns a list of map roots and profiles at each revision.
+func (c *Client) PaginateHistory(ctx context.Context, appID, userID string, start, end int64) (map[uint64]*types.MapRootV1, map[uint64][]byte, error) {
 	if start < 0 {
 		return nil, nil, fmt.Errorf("start=%v, want >= 0", start)
 	}
 	allRoots := make(map[uint64]*types.MapRootV1)
 	allProfiles := make(map[uint64][]byte)
-	epochsReceived := int64(0)
 	epochsWant := end - start + 1
-	for epochsReceived < epochsWant {
-		count := min(int32((end-start)+1), pageSize)
+	for int64(len(allProfiles)) < epochsWant {
+		log.Printf("pageSize: %v", c.pageSize)
+		count := min(int32(epochsWant-int64(len(allProfiles))), c.pageSize)
```

Comment:
Hm, this bit is just as problematic as the previous version due to shrinking cast. Probably making a `min` function for `int64` is too much hassle. Instead, how about rewinding this back, and simply checking `end-start+1` for an overflow before casting it down to `int32` in line 175?

Reply:
Making `min` accept `int64` is actually the better solution. It defers the downcast till the end when we expect the value to be small.

Reply:
Oh, I agree. Wasn't aware that the `min` function was used only here.